garage/contrib/exp/experiment.py

```diff
@@ -0,0 +1,58 @@
+import gym
+
+from garage.contrib.exp import Agent
+from garage.contrib.exp import Checkpointer
+from garage.contrib.exp import Logger
+from garage.contrib.exp import Sampler
+
+
+class Experiment():
+    def __init__(
+            self,
+            env: gym.Env,
+            agent: Agent,
+        checkpointer: Checkpointer,
+            logger: Logger,
+        sampler: Sampler = None,
+        # common experiment parameters:
+            n_itr=10,
+            batch_size=1000,
+    ):
+        if sampler is None:
+            from garage.contrib.exp.samplers import BatchSampler
+            sampler = BatchSampler(env=env)
+
+        if checkpointer.resume:
+            checkpoint = checkpointer.load(sampler=sampler, agent=agent)
+            sampler = checkpoint['sampler']
+            agent = checkpoint['agent']
+
+        self.agent = agent
+        self.checkpointer = checkpointer
+        self.logger = logger
+        self.sampler = sampler
+        self.n_itr = n_itr
+        self.batch_size = batch_size
+
+        self.itr = 0
+
+    def train_once(self):
+        self.itr = self.itr + 1
+        print('Itration', self.itr)
+
+        obs = self.sampler.reset()
+        while self.sampler.sample_count < self.batch_size:
+            actions = self.agent.get_actions(obs)
+            obs = self.sampler.step(actions)
```

Comment:
hmmm the sampler is _not necessarily_ stepping. for instance it might just return some samples from a dataset.

This whole method really confuses me actually. Why does the experiment runner have a method called `train_once()`?

Maybe it should be more like

```python
def train(self):
    # these should probably be constructor parameters, but then we can only accept a sampler Type
    #  in the Experiment constructor. That makes it easier to serialize an experiment, anyway.
    self._sampler.set_agent(self._agent)  
    self._sampler.set_environment(self._environment)
    while self.itr < self.n_itr:
        samples = self.sampler.get_samples()  # sampler already has agent and environment
        self._agent.train_once(samples)
        self._sampler.update_agent(self._agent)

        # Logging
        # TODO: how does this work Peter's logger if the agent has access to the Logger? The agent only has a facade logger, right? Then what type does get_summary() produce?
        self._logger.log(self._agent.get_summary())
        self._logger.log(self._sampler.get_summary())
        self._logger.dumps()

        # Checkpointing
        self._checkpointer.maybe_save(self)  # ??? save the whole experiment
```